api/src/auth/drivers/saml.ts

```diff
@@ -0,0 +1,171 @@
+		const userID = await this.fetchUserID(identifier);
+
+		const userInfo = {
+			name: payload[nameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name'],
+			given_name: payload[givenNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname'],
+			family_name: payload[familyNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname'],
+			email,
+		};
+
+		if (userID) return userID;
+
+		// Is public registration allowed?
+		if (!allowPublicRegistration) {
+			logger.trace(`[Saml] User doesn't exist, and public registration not allowed for provider "${provider}"`);
+			throw new InvalidCredentialsException();
+		}
+
+		try {
+			await this.usersService.createOne({
+				provider,
+				first_name: userInfo.given_name,
+				last_name: userInfo.family_name,
+				email: email,
+				external_identifier: identifier,
+				role: this.config.defaultRoleId,
+				//auth_data: tokenSet.refresh_token && JSON.stringify({ refreshToken: tokenSet.refresh_token }),
+			});
+			const userID = await this.fetchUserID(identifier);
+			if (userID) return userID;
+		} catch (e) {
+			if (e instanceof RecordNotUniqueException) {
+				logger.warn(e, '[Saml] Failed to register user. User not unique');
+				throw new InvalidProviderException();
+			}
+			throw e;
+		}
+
+		throw new InvalidCredentialsException();
+	}
+
+	async login(user: User): Promise<void> {
+		return this.refresh(user);
+	}
+
+	async refresh(user: User): Promise<void> {
+		let authData = user.auth_data as AuthData;
+
+		if (typeof authData === 'string') {
+			try {
+				authData = parseJSON(authData);
+			} catch {
+				logger.warn(`[Saml] Session data isn't valid JSON: ${authData}`);
+			}
+		}
+
+		if (authData?.refreshToken) {
+			try {
+				// TODO
+				/*
+				const tokenSet = await client.refresh(authData.refreshToken);
+				// Update user refreshToken if provided
+				if (tokenSet.refresh_token) {
+					await this.usersService.updateOne(user.id, {
+						auth_data: JSON.stringify({ refreshToken: tokenSet.refresh_token }),
+					});
+				}
+*/
+			} catch (e) {
+				throw handleError(e);
+			}
+		}
+	}
+}
+
+const handleError = (e: any) => {
+	logger.trace(e, `[Saml] Unknown error`);
+	return e;
+};
+
+export function createSAMLAuthRouter(providerName: string) {
+	const router = Router();
+
+	router.get(
+		'/metadata',
+		asyncHandler(async (req, res, next) => {
+			const { sp } = getAuthProvider(providerName) as SAMLAuthDriver;
+			return res.header('Content-Type', 'text/xml').send(sp.getMetadata());
+		})
+	);
+
+	router.get(
+		'/',
+		asyncHandler(async (req, res) => {
+			const { sp, idp } = getAuthProvider(providerName) as SAMLAuthDriver;
+			const { context } = await sp.createLoginRequest(idp, 'redirect');
+			return res.redirect(context);
+		})
+	);
+
+	router.post(
+		'/acs',
+		express.urlencoded({ extended: false }),
+		asyncHandler(async (req, res, next) => {
+			const { sp, idp } = getAuthProvider(providerName) as SAMLAuthDriver;
```

Comment:
This is more of a preference than a problem, but I prefer keeping the underlying client private and exposing functionality through methods. It encourages keeping code within the provider, and keeps things easier to follow. E.g. `const attributes = await provider.parseLoginResponse(req);`.

Either way, you might also want to consider wrapping this in an exception handler.